pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java

```diff
@@ -133,183 +120,143 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
-            _indexCreatorProvider.newTextIndexCreator(context.forTextIndex(fstType, true,
-                TextIndexUtils.extractStopWordsInclude(columnName, _columnProperties),
-                TextIndexUtils.extractStopWordsExclude(columnName, _columnProperties))));
       }
+      _creatorsByColAndIndex.put(columnName, creatorsByIndex);
+    }
 
-      if (fstIndexColumns.contains(columnName)) {
-        _fstIndexCreatorMap.put(columnName, _indexCreatorProvider.newTextIndexCreator(
-            context.forFSTIndex(_config.getFSTIndexType(),
-                (String[]) columnIndexCreationInfo.getSortedUniqueElementsArray())));
+    // Although NullValueVector is implemented as an index, it needs to be treated in a different way than other indexes
+    _nullHandlingEnabled = _config.isNullHandlingEnabled();
+    if (_nullHandlingEnabled) {
+      for (FieldSpec fieldSpec : schema.getAllFieldSpecs()) {
+        // Initialize Null value vector map
+        String columnName = fieldSpec.getName();
+        _nullValueVectorCreatorMap.put(columnName, new NullValueVectorCreator(_indexDir, columnName));
       }
+    }
+  }
 
-      JsonIndexConfig jsonIndexConfig = jsonIndexConfigs.get(columnName);
-      if (jsonIndexConfig != null) {
-        _jsonIndexCreatorMap.put(columnName,
-            _indexCreatorProvider.newJsonIndexCreator(context.forJsonIndex(jsonIndexConfig)));
-      }
+  private FieldIndexConfigs adaptConfig(String columnName, FieldIndexConfigs config,
+      ColumnIndexCreationInfo columnIndexCreationInfo, SegmentGeneratorConfig segmentCreationSpec) {
+    FieldIndexConfigs.Builder builder = new FieldIndexConfigs.Builder(config);
+    // Sorted columns treat the 'forwardIndexDisabled' flag as a no-op
+    ForwardIndexConfig fwdConfig = config.getConfig(StandardIndexes.forward());
+    if (!fwdConfig.isEnabled() && columnIndexCreationInfo.isSorted()) {
+      builder.add(StandardIndexes.forward(), new ForwardIndexConfig.Builder(fwdConfig)
+          .withLegacyProperties(segmentCreationSpec.getColumnProperties(), columnName)
+          .build());
+    }
+    // Initialize inverted index creator; skip creating inverted index if sorted
+    if (columnIndexCreationInfo.isSorted()) {
+      builder.undeclare(StandardIndexes.inverted());
+    }
+    return builder.build();
+  }
 
-      H3IndexConfig h3IndexConfig = h3IndexConfigs.get(columnName);
-      if (h3IndexConfig != null) {
-        _h3IndexCreatorMap.put(columnName,
-            _indexCreatorProvider.newGeoSpatialIndexCreator(context.forGeospatialIndex(h3IndexConfig)));
-      }
+  /**
+   * Returns true if the given index type has their own construction lifecycle and therefore should not be instantiated
+   * in the general index loop and shouldn't be notified of each new column.
+   */
+  private boolean hasSpecialLifecycle(IndexType<?, ?, ?> indexType) {
+    return indexType == StandardIndexes.nullValueVector() || indexType == StandardIndexes.dictionary();
```

Comment:
Can this be a property on the indexType itself? So the call can be indexType.hasSpecialLifecycle()?

Reply:
I don't think that is a good idea because `SegmentColumnarIndexCreator` has to know not only which indexes have to be treated "in a special way" but also _how_ to treat them in a special way. I mean, imagine that a new index type has to be treated in a special way. It wouldn't be enough to just return true to some new method, `SegmentColumnarIndexCreator` would need to know how to deal with that index.

We can design a more complex algorithm where there are different strategies, but it sounds difficult to find the correct abstraction and TBH I don't think new indexes will need to be special. In fact dictionary and null vector are not actual indexes. They are different ways to encode the information that are historically implemented as indexes because that was the tool we had.